### PR TITLE
Eliminate layer of callables by making _Asset a non-callable NamedTuple

### PR DESCRIPTION
**Pull request by schrockn**
## Summary & Motivation

This eliminates a bunch of necessary assignments and reduces an indent level by making `_Asset` a plain named tuple and its `__call__` function a normal function

## How I Tested These Changes

BK

### CODE REVIEW

**Comment by schrockn** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22235?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22235** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22235?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22234** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22232** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22232?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22230** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22229** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22228** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22227** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22225** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22225?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#22226](https://github.com/dagster-io/dagster/pull/22226) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22226?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#22224** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22223** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#22222** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @schrockn and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py [hunk 1]
```diff
@@ -306,184 +306,153 @@ def resolve_asset_key_and_name_for_decorator(
...
-        self.owners = owners
-
-    def __call__(self, fn: Callable[..., Any]) -> AssetsDefinition:
-        from dagster._config.pythonic_config import (
-            validate_resource_annotated_function,
-        )
-        from dagster._core.execution.build_resources import wrap_resources_for_execution
-
-        validate_resource_annotated_function(fn)
+class AssetArgs(NamedTuple):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:309]:
```suggestion
class AssetDecoratorArgs(NamedTuple):
```

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py [hunk 2]
```diff
@@ -306,184 +306,153 @@ def resolve_asset_key_and_name_for_decorator(
...
+    auto_materialize_policy: Optional[AutoMaterializePolicy]
+    backfill_policy: Optional[BackfillPolicy]
+    retry_policy: Optional[RetryPolicy]
+    code_version: Optional[str]
+    key: Optional[CoercibleToAssetKey]
+    check_specs: Optional[Sequence[AssetCheckSpec]]
+    owners: Optional[Sequence[str]]
+
+
+def invoke(args: AssetArgs, fn: Callable[..., Any]) -> AssetsDefinition:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:338]: I'm having a little trouble connecting the name of this function to what it does. Without more context about what's going on, I would expect a function named `invoke` that accepts a `Callable` to invoke that callable.

Thoughts on something verbose like `create_assets_def_from_fn_and_decorator_args`?

**Review by sryza — CHANGES_REQUESTED** [review]

File: python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py (hunk 2 above)
Comment (schrockn) [reply python_modules/dagster/dagster/_core/definitions/decorators/asset_decorator.py:329]: For good reason because `invoke` is a terrible name

**Review by sryza — APPROVED** [review]

**Comment by schrockn** [comment]
### Merge activity

* **Jun 7, 3:53 AM EDT**: @schrockn started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22235).

**Pull request merged**